src/click/types.py

```diff
@@ -107,6 +107,19 @@ def fail(self, message, param=None, ctx=None):
         """Helper method to fail with an invalid value message."""
         raise BadParameter(message, ctx=ctx, param=param)
 
+    def shell_complete(self, ctx, args, incomplete):
+        """Return a list of completions for the incomplete value. Most
+        types do not provide completions, but some do, and this allows
+        custom types to provide custom completions as well.
+
+        :param ctx: Invocation context for this command.
+        :param args: List of complete args before the incomplete value.
+        :param incomplete: Value being completed. May be empty.
+
```

Comment:
:x: i would love to have `:rtype:` documented here as well.

Reply:
We'll be adding typing annotations to the whole library eventually.